src/Agent/NewRelic/Agent/Core/Attributes/LogContextDataFilter.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright 2020 New Relic, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using NewRelic.Agent.Configuration;
+
+namespace NewRelic.Agent.Core.Attributes
+{
+    public interface ILogContextDataFilter
+    {
+        Dictionary<string, object> FilterLogContextData(Dictionary<string, object> contextData);
+    }
+
+    public class LogContextDataFilter : ILogContextDataFilter
+    {
+        private IConfigurationService _configurationService;
+        private List<LogContextDataFilterRule> _includeRuleList;
+        private List<LogContextDataFilterRule> _excludeRuleList;
+        private List<LogContextDataFilterRule> _orderedCludeRuleList;
+
+        private Dictionary<string, bool> _clusionCache = new Dictionary<string, bool>();
+
+        public LogContextDataFilter(IConfigurationService configurationService)
+        {
+            _configurationService = configurationService;
+        }
+
+
+        public Dictionary<string, object> FilterLogContextData(Dictionary<string, object> unfilteredContextData)
+        {
+            if (unfilteredContextData == null)
+            {
+                return null;
+            }
+
+            var filteredContextData = new Dictionary<string, object>();
+
+            // Pseudocode
+            // Include/exclude rules:
+            // 1. Exclude wins over include
+            // 2. Can have a '*' wildcard at the end of an include/exclude definition
+            // 3. "more specific" rules win over "less specific".  Example: include=AB, exclude=A*, AB would be included but not ABC or AAA
+
+            // Algorithm:
+            // Create list of rules, sorted by most specific to least specific, with excludes first and then includes within each level of specificity
+            // Specificity is just the length of the string, minus any wildcard character at the end
+            // For each context data key/value pair:
+            //   First, check the "clusion cache" to see if we already know whether or not to include or exclude this key, and if so handle it appropriately
+            //   Otherwise, go through the list of rules and see if we find a matching rule to exclude or include, cache the result, and handle it appropriately
+
+            foreach (var kvp in unfilteredContextData)
+            {
+                if (!_clusionCache.ContainsKey(kvp.Key))
```

Comment:
_clusionCache should be reset during configuration change events as well

Reply:
Done